api/proto/teleport/legacy/types/types.proto

```diff
@@ -2453,6 +2453,19 @@ enum CreateHostUserMode {
   HOST_USER_MODE_KEEP = 3;
 }
 
+// CreateDatabaseUserMode determines whether database user creation should be
+// disabled or if users should be cleaned up or kept after sessions end.
+enum CreateDatabaseUserMode {
+  DB_USER_MODE_UNSPECIFIED = 0;
+  // DB_USER_MODE_OFF disables user creation.
+  DB_USER_MODE_OFF = 1;
+  // DB_USER_MODE_KEEP allows user creation and disable users at session end.
+  DB_USER_MODE_KEEP = 2;
+  // DB_USER_MODE_PREFER_DROP allows user creation and prefers dropping users
+  // if they don't own any resources. Otherwise, it disables them.
+  DB_USER_MODE_PREFER_DROP = 3;
```

Comment:
My thought is to make these modes work the same way on each engine, making it easier for users to understand what is going to be the effect of each mode. This could lead to modes not making much sense in some engines, so you might be right about it being specific.

Note: This is not necessarily a limitation on the PostgreSQL side. It is just more trivial than other engines. We could use the `DROP OWNED` to drop the user objects or call the `REASSIGN OWNED` to change their owner.

> MySQL and MongoDB don't have ownership so users can be dropped without any issue.

Could a user create a table that only they can access? Or do their roles guarantee other users access to it?

I wonder if there could be any advantage to using the "prefer drop" for them.

Reply:
best_effort_drop?

Reply:
> Could a user create a table that only they can access?

No. The table is owned by the database, not the user who produces it (communism?).

> My thought is to make these modes work the same way on each engine, making it easier for users to understand what is going to be the effect of each mode. This could lead to modes not making much sense in some engines, so you might be right about it being specific.

Ok. So will MySQL support a separate `drop` mode besides `preferDrop`?

Reply:
> No. The table is owned by the database, not the user who produces it (communism?).

I see. However, there is a case where there can be a problem: when the user creates procedures. The drop command only returns a warning, but when anyone calls the procedure, it returns an error. (This also applies to views).

The point is: to make a safe version of the drop command, and we could discuss having another version that could drop users anyway (or execute the necessary commands to ensure it will succeed).

> Ok. So will MySQL support a separate drop mode besides preferDrop?

If we think a `drop` makes sense, yes. Then all the engines would need to implement it also.

Reply:
> best_effort_drop?

That sounds like a better option. And since we already have this "format" on other parts, it would be better to keep it.